### PR TITLE
Laboratory work #2, Ratmir Akhmatzyanov - 22FPL1

### PR DESCRIPTION
**Pull request by Rotmerka**

### CODE REVIEW

**Comment by artyomtugaryov** [comment]
Nothing to review

**Comment by Rotmerka** [comment]
Возможно ли как-то запускать тесты не делая каждый раз коммит?

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +14,14 @@ def prepare_word(
...
     :return: preprocessed word
     """
+    if not (
+        isinstance(raw_word, str)
+        and isinstance(start_of_word, str | None)
+        and isinstance(end_of_word, str | None)
+    ):
+        return None
+
+    return tuple(filter(None, [start_of_word, *(i for i in raw_word), end_of_word]))
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:24]: Great solution

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -36,6 +56,16 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_freq = {}
+    for word in word_frequencies.keys():
+        pair_freq = {(word[i], word[i + 1]): word_frequencies[word] for i in range(len(word) - 1)}
+    #    for i in range(len(word) - 1):
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:65]: ?

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -36,6 +56,16 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_freq = {}
+    for word in word_frequencies.keys():
+        pair_freq = {(word[i], word[i + 1]): word_frequencies[word] for i in range(len(word) - 1)}
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:64]: You create a new dict in each iteration

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (Rotmerka) [reply lab_2_tokenize_by_bpe/main.py:65]: At this point I hashed the working code to see if i could make it through a dictionary generator

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (Rotmerka) [reply lab_2_tokenize_by_bpe/main.py:24]: Thank you!

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (Rotmerka) [reply lab_2_tokenize_by_bpe/main.py]: Fixed

**Pull request closed without merging**